### PR TITLE
Upgrade types and bundles as per bundle manifest headers

### PR DESCRIPTION
**Pull request by ahgittin**
Follow on from #872 which actually applies the upgrade in most places.

One known gap, and possibly some unknown gaps, but the main cases are working.

Suggest revisiting post-YOML as that will make it easier to handle upgrades due to a single instantiation path (instead of the many we currently have).

### CODE REVIEW

**Comment by ahgittin** [comment]
retest this please

non-det failure:

```
Error Message

expected one element but was: <{myKey=myDefault}, {myKey=valIsV1}>
Stacktrace

java.lang.IllegalArgumentException: expected one element but was: <{myKey=myDefault}, {myKey=valIsV1}>
	at org.apache.brooklyn.enricher.stock.UpdatingMapTest.testUpdateMapEmitsEventOnChange(UpdatingMapTest.java:126)
Standard Output

2017-11-01 17:53:23,945 INFO  TESTNG INVOKING CONFIGURATION: "Surefire test" - @AfterMethod org.apache.brooklyn.core.test.BrooklynMgmtUnitTestSupport.tearDown()
2017-11-01 17:53:23,947 INFO  Stopping application BasicApplicationImpl{id=o3zm3t4g8w}
2017-11-01 17:53:23,952 INFO  Stopped application BasicApplicationImpl{id=o3zm3t4g8w}
2017-11-01 17:53:23,954 INFO  TESTNG PASSED CONFIGURATION: "Surefire test" - @AfterMethod org.apache.brooklyn.core.test.BrooklynMgmtUnitTestSupport.tearDown() finished in 9 ms
2017-11-01 17:53:23,954 INFO  TESTNG INVOKING CONFIGURATION: "Surefire test" - @BeforeMethod org.apache.brooklyn.core.test.BrooklynAppUnitTestSupport.setUp()
2017-11-01 17:53:23,954 INFO  Added external config supplier named 'brooklyn-demo-sample': org.apache.brooklyn.core.config.external.InPlaceExternalConfigSupplier@715fceaf
2017-11-01 17:53:23,963 INFO  TESTNG PASSED CONFIGURATION: "Surefire test" - @BeforeMethod org.apache.brooklyn.core.test.BrooklynAppUnitTestSupport.setUp() finished in 9 ms
2017-11-01 17:53:23,963 INFO  TESTNG INVOKING: "Surefire test" - org.apache.brooklyn.enricher.stock.UpdatingMapTest.testUpdateMapEmitsEventOnChange()
2017-11-01 17:53:23,979 INFO  TESTNG FAILED: "Surefire test" - org.apache.brooklyn.enricher.stock.UpdatingMapTest.testUpdateMapEmitsEventOnChange() finished in 15 ms
java.lang.IllegalArgumentException: expected one element but was: <{myKey=myDefault}, {myKey=valIsV1}>
	at com.google.common.collect.Iterators.getOnlyElement(Iterators.java:317)
	at com.google.common.collect.Iterables.getOnlyElement(Iterables.java:289)
	at org.apache.brooklyn.enricher.stock.UpdatingMapTest.testUpdateMapEmitsEventOnChange(UpdatingMapTest.java:126)
```

**Comment by ahgittin** [comment]
Have patched the known gap around entity spec.  Those tests now passing.

That required a surprising series of tweaks:  to how EntitySpec is resolved/persisted, to how default values are coerced/resolved, and to which iterables are allowed to be traversed for deep resolution.  All changes I think are good independent of each other - it just so happened the first exposed the second etc.

---


So now not aware of anything which isn't upgraded as far as we can.  It is still a problem that we persist the _java_ type names rather than the _registered type_ names, so changes in an upgraded registered type definition are not actually picked up -- but at least we record the correct type name and look for java in the correct bundle.  And many of the places we want to change to work with RT instead of JT are now easy to find for when we do that change (and tests are aware of change semantics) by modifying and searching for uses of:

    CatalogUpgrades.markerForCodeThatLoadsJavaTypesButShouldLoadRegisteredType()

**Comment by robertgmoss** [comment]
I have tested rebinding using the [brooklyn-etcd](https://github.com/brooklyncentral/brooklyn-etcd) entity, where the following was added to its `pom.xml`:

```           
            <plugin>
                <groupId>org.apache.felix</groupId>
                <artifactId>maven-bundle-plugin</artifactId>
                <extensions>true</extensions>
                <configuration>
                    <instructions>
                        <Brooklyn-Catalog-Force-Remove-Bundles>*</Brooklyn-Catalog-Force-Remove-Bundles>
                        <Brooklyn-Catalog-Upgrade-For-Bundles>*</Brooklyn-Catalog-Upgrade-For-Bundles>
                    </instructions>
                </configuration>
                <executions>
                    <execution>
                        <id>bundle-manifest</id>
                        <phase>process-classes</phase>
                        <goals>
                            <goal>manifest</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
```

and this appears to work.

File: launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java [hunk 1]
```diff
@@ -740,20 +744,82 @@ public void testRebindUpgradeEntity() throws Exception {
...
+        if (CatalogUpgrades.markerForCodeThatLoadsJavaTypesButShouldLoadRegisteredType()) {
+            Assert.assertEquals(Entities.deproxy(entity).getClass().getName(), BasicEntityImpl.class.getName());
+        } else {
+            Assert.assertEquals(Entities.deproxy(entity).getClass().getName(), "com.example.brooklyn.test.osgi.entities.SimpleEntityImpl");
+        }
+    }
+    
+    @Test
+    public void testRebindUpgradeReferencedEntityFromCatalogAndDeployment() throws Exception {
+        File initialBomFileV2 = prepForRebindRemovedItemTestReturningBomV2(false, true);
```
Comment (aledsage) [inline launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java]: I presume this should pass `CatalogUpgrades.markerForCodeThatLoadsJavaTypesButShouldLoadRegisteredType()` as the first arg, given that the `startupAssertions` will adjust its assertions based on that.

**Comment by robertgmoss** [comment]
LGTM

File: core/src/main/java/org/apache/brooklyn/core/mgmt/persist/XmlMementoSerializer.java [hunk 2]
```diff
@@ -341,7 +371,7 @@ public void marshal(Object source, HierarchicalStreamWriter writer, MarshallingC
         public Object unmarshal(HierarchicalStreamReader reader, UnmarshallingContext context) {
             if (reader.hasMoreChildren()) {
                 Class<?> type = readClassType(reader, mapper);
-//                Class<?> type2 = context.getRequiredType();
+                // could confirm it is subtype of context.getRequiredType()
```
[the remark this answers is not in the document]
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/mgmt/persist/XmlMementoSerializer.java]: TODO to me implies a bit stronger intention, here it's a "not yet sure should we do"

File: core/src/main/java/org/apache/brooklyn/core/typereg/BundleUpgradeParser.java [hunk 3]
```diff
@@ -312,6 +335,90 @@ public static boolean contains(Iterable<VersionRangedName> names, VersionedName
...
+            return false;
+        }
+        
+        @Beta
+        CatalogUpgrades withTypeCleared(VersionedName versionedName) {
+            return builder().addAll(this).clearUpgradesProvidedByType(versionedName).build();
+        }
+        @Beta
+        CatalogUpgrades withBundleCleared(VersionedName versionedName) {
+            return builder().addAll(this).clearUpgradesProvidedByType(versionedName).build();
```
[the remark this answers is not in the document]
Comment (ahgittin) [reply core/src/main/java/org/apache/brooklyn/core/typereg/BundleUpgradeParser.java]: yes!

File: launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java (hunk 1 above)
Comment (ahgittin) [reply launcher/src/test/java/org/apache/brooklyn/launcher/BrooklynLauncherRebindCatalogOsgiTest.java]: yes

**Comment by ahgittin** [comment]
Thanks @aledsage @robertgmoss - comments addressed and merged

**Pull request merged**